src/Features/Core/Diagnostics/AbstractHostDiagnosticUpdateSource.cs

```diff
@@ -74,10 +76,16 @@ internal void ReportAnalyzerDiagnostic(DiagnosticAnalyzer analyzer, Diagnostic d
 
         public void ClearAnalyzerReferenceDiagnostics(AnalyzerFileReference analyzerReference, string language, ProjectId projectId)
         {
-            foreach (var analyzer in analyzerReference.GetAnalyzers(language))
+            var analyzers = analyzerReference.GetAnalyzers(language);
+            ClearAnalyzerDiagnostics(analyzers, projectId);
+            AnalyzerManager.Instance.ClearCachedAnalyzerDescriptors(analyzers);
```

Comment:
I am going to do this refactoring as part of a separate change though, its not mandatory for this fix and I want to get this done asap.